timeseries/src/main/scala/com/criteo/cuttle/timeseries/TimeSeriesScheduler.scala

```diff
@@ -960,12 +988,46 @@ case class TimeSeriesScheduler(logger: Logger,
         .whenIsDef(dependenciesSatisfied)
         .whenIsDef(noChildrenRunning)
 
-      for {
+      val job2Contexts = for {
         (interval, maybeBackfill) <- toRun.toList
-        (lo, hi) <- job.scheduling.calendar.inInterval(interval, job.scheduling.maxPeriods)
+        (lo, hi) <- job.scheduling.calendar.inInterval(interval, job.scheduling.batching.size)
       } yield {
+        logger.debug(s"creating a context for ${job.name} for interval $lo, $hi")
         (job, TimeSeriesContext(lo, hi, maybeBackfill, projectVersion))
       }
+
+      if (job2Contexts.isEmpty) {
+        logger.debug(s"nothing to run for ${job.name}")
+        atomic { implicit txn =>
+          debouncedJobs() = debouncedJobs() - job
+        }
+        List.empty
+      } else {
+        if (job.scheduling.batching.size > 1) {
+          val now = Instant.now()
+          debouncedJobs.single().contains(job) match {
```

Comment:
- There is no concurrent access to _debouncedJobs_ (scheduling loops are serialized), so technically we do not need STM
- No side effect in _jobsToRun_, please take as an argument & return the _debouncedJobs_ from _jobsToRun_. You can side effect them in the calling method where all the rest is side-effected